### ISSUE TITLE
What does it mean for instrumentation to support Samplers that return SamplingDecision.DROP?

### ISSUE THREAD

**Issue by trask** [open-telemetry/opentelemetry-java-instrumentation#2572]
#### Example 1

If a Sampler returns `SamplingDecision.DROP` only for B below, should C's parent be A?

    Server Span (A) -> Internal Span (B) -> Client Span (C)

#### Example 2

If a Sampler returns `SamplingDecision.DROP` only for C below, should D's parent be B?

    Server Span (A) -> Client Span (B) -> Server Span (C) -> Client Span (D)

#### Initial research

The [SamplingDecision javadoc](https://javadoc.io/doc/io.opentelemetry/opentelemetry-sdk-trace/latest/io/opentelemetry/sdk/trace/samplers/SamplingDecision.html#DROP) says:

* DROP

  > Span is dropped. The resulting span will be completely no-op.

* RECORD_ONLY

  > Span is recorded only. The resulting span will record all information like timings and
     attributes but will not be exported. Downstream {@linkplain Sampler#parentBased(Sampler)
     parent-based} samplers will not sample the span.

Note that unlike `RECORD_ONLY`, `DROP` does not say that downstream parent-based samplers will not sample the span.

Which may mean that a "dropped" span should not be used as a parent, because otherwise it would cause parent-based samplers to not sample downstream spans.

"completely no-op" also seems to point to not propagating or using "dropped" spans as a parent, and allowing the existing active span to stay active.

But really I have no idea. I could not find anything about this particular issue. Is this even a thing Samplers are allowed to do, return `DROP` for just one span in a whole trace?

cc: @bogdandrutu

**Comment by anuraaga** [comment]
tl;dr I suspect `RECORD_ONLY` has the text because it's ambiguous, while it's the same for `DROP` - downstream parent-based sampling won't happen.

I think a sampler can do anything it wants, including dropping a single span in the middle of a tree (rare naturally). This should be respected by the next sampling decision in the tree, so a parent-based sampler would also drop `C`, or `D` in those examples. If it's not a parent-based sampler, `C` and `D` may be sampled anyways, it depends on the sampler implementation.

**Comment by trask** [comment]
> If it's not a parent-based sampler, C and D may be sampled anyways, it depends on the sampler implementation.

Do you have thoughts on what their parents should be when we export them? Should their parents be the "dropped" span (I believe this is the current javaagent behavior)?

**Comment by anuraaga** [comment]
@trask I'd expect it to have the parent of the dropped span, which most backends will render as a broken trace as we'd sort of expect. It allows essentially "restarting a trace" for debugging things downstream. It does mean there seems to be no way for `C` or `D` to "start a new trace" via the `Sampler` but I guess there's no real use case for that.

**Comment by trask** [comment]
Thanks, this makes sense to me 👍
